src/Core/Checkout/Promotion/Subscriber/PromotionIndividualCodeRedeemer.php

```diff
@@ -25,21 +32,54 @@ class PromotionIndividualCodeRedeemer implements EventSubscriberInterface
      * @internal
      *
      * @param EntityRepository<PromotionIndividualCodeCollection> $codesRepository
+     * @param EntityRepository<OrderCollection> $orderRepository
      */
-    public function __construct(private readonly EntityRepository $codesRepository)
-    {
+    public function __construct(
+        private readonly EntityRepository $codesRepository,
+        private readonly EntityRepository $orderRepository
+    ) {
     }
 
     public static function getSubscribedEvents(): array
     {
         return [
-            CheckoutOrderPlacedEvent::class => 'onOrderPlaced',
+            CheckoutOrderPlacedEvent::class => 'onOrderProcess',
+            OrderEvents::ORDER_LINE_ITEM_WRITTEN_EVENT => 'onOrderLineItemWritten',
         ];
     }
 
-    public function onOrderPlaced(CheckoutOrderPlacedEvent $event): void
+    public function onOrderLineItemWritten(EntityWrittenEvent $event): void
+    {
+        if ($event->getContext()->getVersionId() !== Defaults::LIVE_VERSION) {
+            return;
+        }
+
+        $orderLineItems = new OrderLineItemCollection();
+
+        foreach ($event->getWriteResults() as $result) {
+            $orderLineItems->add((new OrderLineItemEntity())->assign($result->getPayload()));
+        }
+
+        $lineItem = $orderLineItems->first();
+        \assert($lineItem instanceof OrderLineItemEntity);
+
+        $order = $this->orderRepository->search(
+            new Criteria([$lineItem->getOrderId()]),
+            $event->getContext()
+        )->getEntities()->first();
+        \assert($order instanceof OrderEntity);
+
+        $this->redeemCode($orderLineItems, $order, $event->getContext());
+    }
+
+    public function onOrderProcess(CheckoutOrderPlacedEvent $event): void
+    {
+        $this->redeemCode($event->getOrder()->getLineItems(), $event->getOrder(), $event->getContext());
+    }
```

Comment:
why do we still need action on `CheckoutOrderPlaced`? There, line items are written as well, right?